Reader.go

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2014 by Farsight Security, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package framestream
+
+import (
+	"bufio"
+	"encoding/binary"
+	"io"
+	"io/ioutil"
+	"log"
+	"time"
+)
+
+type ReaderOptions struct {
+	// The ContentTypes accepted by the Reader. May be left unset for no
+	// content negotiation. If the corresponding Writer offers a disjoint
+	// set of ContentTypes, NewReader() will return ErrContentTypeMismatch.
+	ContentTypes [][]byte
+	// If Bidirectional is true, the underlying io.Reader must be an
+	// io.ReadWriter, and the Reader will engage in a bidirectional
+	// handshake with its peer to establish content type and communicate
+	// shutdown.
+	Bidirectional bool
+	// Timeout gives the timeout for reading the initial handshake messages
+	// from the peer and writing response messages if Bidirectional. It is
+	// only effective for underlying Readers satisfying net.Conn.
+	Timeout time.Duration
+}
+
+// Reader reads data frames from an underlying io.Reader using the Frame
+// Streams framing protocol.
+type Reader struct {
+	contentType   []byte
+	bidirectional bool
+	r             *bufio.Reader
+	w             *bufio.Writer
+	stopped       bool
+}
+
+// NewReader creates a Frame Streams Reader reading from the given io.Reader
+// with the given ReaderOptions.
+func NewReader(r io.Reader, opt *ReaderOptions) (*Reader, error) {
+	if opt == nil {
+		opt = &ReaderOptions{}
+	}
+	tr := timeoutReader(r, opt)
+	reader := &Reader{
+		bidirectional: opt.Bidirectional,
+		r:             bufio.NewReader(tr),
+		w:             nil,
+	}
+
+	if opt.ContentTypes != nil {
+		reader.contentType = opt.ContentTypes[0]
+	}
+
+	var cf ControlFrame
+	if opt.Bidirectional {
+		w, ok := tr.(io.Writer)
+		if !ok {
+			return nil, ErrType
+		}
+		reader.w = bufio.NewWriter(w)
+
+		// Read the ready control frame.
+		err := cf.DecodeTypeEscape(reader.r, CONTROL_READY)
```

Comment:
BTW, take a look at ControlFrame.Decode(), which DecodeTypeEscape() calls. Shouldn't we put a limit on the cflen value, which is read from the network? Currently it can take any uint32 value, which is used later in `make([]byte, int(cflen))`.

Reply:
Agreed, this should probably be limited to something lower than 4GB :)

FWIW, the C frame streams code seems to impose a 512 byte limit on the length of a control frame.

Reply:
512B sounds reasonable. After this correction, I'm fine with this pull request and I think it should be merged.